### ISSUE TITLE
Missing Methods (Possibly AOT Trimmed)

### ISSUE THREAD

**Issue by stephnux** [dotnet/efcore#35105]
It seems the following functions disappeared from Microsoft.EntityFrameworkCore.Relational 9.0.0 on .net 9.0

```C#
RelationalQueryableExtensions.ExecuteDeleteAsync
RelationalQueryableExtensions.ExecuteDelete
RelationalQueryableExtensions.ExecuteUpdateAsync
RelationalQueryableExtensions.ExecuteUpdate
```

This does not appear in the list of breaking changes.
[EF-Core-9.0 Breaking Changes](https://learn.microsoft.com/en-us/ef/core/what-is-new/ef-core-9.0/breaking-changes)

## See the stack trace below:
```
[11:41:15 Fatal Duende.IdentityServer.Hosting.IdentityServerMiddleware  12] Unhandled exception: "Method not found: 'System.Threading.Tasks.Task`1<Int32> Microsoft.EntityFrameworkCore.RelationalQueryableExtensions.ExecuteDeleteAsync(System.Linq.IQueryable`1<!!0>, System.Threading.CancellationToken)'."
System.MissingMethodException: Method not found: 'System.Threading.Tasks.Task`1<Int32> Microsoft.EntityFrameworkCore.RelationalQueryableExtensions.ExecuteDeleteAsync(System.Linq.IQueryable`1<!!0>, System.Threading.CancellationToken)'.
   at Duende.IdentityServer.EntityFramework.Stores.PushedAuthorizationRequestStore.ConsumeByHashAsync(String referenceValueHash)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Duende.IdentityServer.EntityFramework.Stores.PushedAuthorizationRequestStore.ConsumeByHashAsync(String referenceValueHash)
   at Duende.IdentityServer.Services.PushedAuthorizationService.ConsumeAsync(String referenceValue) in /_/src/IdentityServer/Services/Default/PushedAuthorizationService.cs:line 35
   at Duende.IdentityServer.Endpoints.Results.AuthorizeHttpWriter.ConsumePushedAuthorizationRequest(AuthorizeResult result) in /_/src/IdentityServer/Endpoints/Results/AuthorizeResult.cs:line 92
   at Duende.IdentityServer.Endpoints.Results.AuthorizeHttpWriter.WriteHttpResponse(AuthorizeResult result, HttpContext context) in /_/src/IdentityServer/Endpoints/Results/AuthorizeResult.cs:line 75
   at Duende.IdentityServer.Endpoints.Results.EndpointResult`1.ExecuteAsync(HttpContext context) in /_/src/IdentityServer/Endpoints/Results/EndPointResult.cs:line 34
   at Duende.IdentityServer.Hosting.IdentityServerMiddleware.Invoke(HttpContext context, IdentityServerOptions options, IEndpointRouter router, IUserSession userSession, IEventService events, IIssuerNameService issuerNameService, ISessionCoordinationService sessionCoordinationService) in /_/src/IdentityServer/Hosting/IdentityServerMiddleware.cs:line 111
[11:41:15 Error Serilog.AspNetCore.RequestLoggingMiddleware  12] HTTP "GET" "/connect/authorize/callback" responded 500 in 1124.1767 ms

```

See the following screenshot from ILSpy

![Image](https://github.com/user-attachments/assets/99752c7f-4bc2-404f-afe4-a98a88a88008)

**Comment by roji** [comment]
@larcai these methods have been moved to EntityFrameworkQueryableExtensions, to allow non-relational usage; that's a binary break (but not a source break).

**Comment by slava-se** [comment]
> [@larcai](https://github.com/larcai) these methods have been moved to EntityFrameworkQueryableExtensions, to allow non-relational usage; that's a binary break (but not a source break).

@roji for me it's a source break. It is especially a problem because the method RelationalQueryableExtensions.ExecuteUpdate is available in DEBUG mode but not in RELEASE mode. I mean, I can type the name explicitly instead of using nameof() but now I'm not sure what should be the DeclaringType.

Still, maybe there is an alternative way to achive my functionality. 
I'm using this extension to set date updated on ExecuteUpdateAsync and I took this code from StackOverflow. I've used it in .NET 8 but now it's not compiling in RELEASE mode. 
Is there a proper/better way to set DateUpdated on Auditable? 
And if you have a bit more time, is there a way to set BrowserInfo on Auditable which is JsonB? (commented part)

```
internal class ExecuteUpdateInterceptor : IQueryExpressionInterceptor
{
    List<(Type Type, Delegate Calls, Func<IEntityType, bool> Filter)> items = new();

    public ExecuteUpdateInterceptor Add<TSource>(
        Func<Expression<Func<SetPropertyCalls<TSource>, SetPropertyCalls<TSource>>>> source,
        Func<IEntityType, bool> filter = null)
    {
        items.Add((typeof(TSource), source, filter));
        return this;
    }

    Expression IQueryExpressionInterceptor.QueryCompilationStarting(
        Expression queryExpression, QueryExpressionEventData eventData)
    {
        if (queryExpression is MethodCallExpression call &&
            call.Method.DeclaringType == typeof(RelationalQueryableExtensions) &&
            call.Method.Name == nameof(RelationalQueryableExtensions.ExecuteUpdate))
        {
            var setPropertyCalls = (LambdaExpression)((UnaryExpression)call.Arguments[1]).Operand;
            var body = setPropertyCalls.Body;
            var parameter = setPropertyCalls.Parameters[0];
            var targetType = eventData.Context?.Model.FindEntityType(parameter.Type.GetGenericArguments()[0]);
            if (targetType != null)
            {
                foreach (var item in items)
                {
                    if (!item.Type.IsAssignableFrom(targetType.ClrType)) continue;
                    if (item.Filter != null && !item.Filter(targetType)) continue;
                    var calls = (LambdaExpression)item.Calls.Method.GetGenericMethodDefinition()
                        .MakeGenericMethod(targetType.ClrType)
                        .Invoke(null, null);
                    body = calls.Body.ReplaceParameter(calls.Parameters[0], body);
                }
                if (body != setPropertyCalls.Body)
                    return call.Update(call.Object, new[] { call.Arguments[0], Expression.Lambda(body, parameter) });
            }
        }
        return queryExpression;
    }
}

public static class ExpressionUtils
{
    public static Expression<Func<SetPropertyCalls<TSource>, SetPropertyCalls<TSource>>> SetAuditable<TSource>() where TSource : Auditable
    {
        var browserInfo = StaticContextHelpers.GetBrowserInfo();
        return s => s.SetProperty(p => p.DateUpdated, DateTime.UtcNow);//.SetProperty(p => p.BrowserInfo, browserInfo);
    }

    public static Expression ReplaceBodyParameter<T, TResult>(this Expression<Func<T, TResult>> source, Expression value)
        => source.Body.ReplaceParameter(source.Parameters[0], value);

    public static Expression ReplaceParameter(this Expression source, ParameterExpression target, Expression replacement)
        => new ParameterReplacer(target, replacement).Visit(source);

    class ParameterReplacer : ExpressionVisitor
    {
        readonly ParameterExpression target;
        readonly Expression replacement;
        public ParameterReplacer(ParameterExpression target, Expression replacement)
            => (this.target, this.replacement) = (target, replacement);
        protected override Expression VisitParameter(ParameterExpression node)
            => node == target ? replacement : node;
    }
}
```

.......

```
builder.Services.AddDbContext<DatabaseContext>
    (
        options =>
        {
            options.UseNpgsql(o => o.UseNetTopologySuite());
            options.UseTriggers(triggerOptions =>
            {
                triggerOptions.AddTrigger<SetDateUpdated>();
            });

            options.AddInterceptors(new ExecuteUpdateInterceptor().Add(ExpressionUtils.SetAuditable<Auditable>));
        }, ServiceLifetime.Transient, ServiceLifetime.Transient
    );
```

**Comment by roji** [comment]
> It is especially a problem because the method RelationalQueryableExtensions.ExecuteUpdate is available in DEBUG mode but not in RELEASE mode.

That doesn't sound right - we didn't vary anything here based on DEBUG vs. RELEASE

> I mean, I can type the name explicitly instead of using nameof() but now I'm not sure what should be the DeclaringType.

I'm not sure I follow - why not just reference EntityFrameworkQueryableExtensions instead of the previous RelationalQueryableExtensions on the following line?

> call.Method.DeclaringType == typeof(RelationalQueryableExtensions) &&

> And if you have a bit more time, is there a way to set BrowserInfo on Auditable which is JsonB? (commented part)

Are you asking whether you can use ExecuteUpdate to change a property inside a `jsonb`-mapped POCO? If so, then unfortunately that's not possible yet... You can use SQL to do that, as a workaround.

**Comment by slava-se** [comment]
> I'm not sure I follow - why not just reference EntityFrameworkQueryableExtensions instead of the previous RelationalQueryableExtensions on the following line?

@roji sorry for wasting your time on that point. Somehow rebuilding the solution didn't fix the problem but VS project clean and then build did.

> Are you asking whether you can use ExecuteUpdate to change a property inside a `jsonb`-mapped POCO? If so, then unfortunately that's not possible yet... You can use SQL to do that, as a workaround.

If ExecuteUpdateAsync is called then I'm trying to overwrite DateUpdated cell (which works with the code above) and also the BrowserInfo which is stored in the JsonB column which doesn't work. From what I understand, it works with DateUpdated because it's a CLR type.
Is it possible to write a custom object on this way as well?

For the classic pattern like 
```
var user = await _dbContext.Users.FirstAsync();
user.Name = "Roji";
await _dbContext.SaveChangesAsync();
```
I'm using EntityFrameworkCore.Triggered library which gets BrowserInfo from the request context and writes it into Auditable.BrowserInfo  but is there something to achieve the same without retrieving data first for ExecuteUpdateAsync() usage?

```
public class User : Auditable
{
    [Key]
    public int Id { get; set; }
    public string Name { get; set; }
}

public abstract class Auditable
{
	[DatabaseGenerated(DatabaseGeneratedOption.Identity)]
	public DateTime DateCreated { get; set; }

	public DateTime DateUpdated { get; set; }

	[Column(TypeName = "jsonb")]
	public BrowserInfo? BrowserInfo { get; set; }
}

public class BrowserInfo
{
    [JsonConverter(typeof(JsonStringEnumConverter))]
    public Browser Browser { get; init; }
    public Version BrowserVersion { get; init; } = null!;
}
```

**Comment by roji** [comment]
@slava-se please post a minimal code sample of what it is that you're trying to do, and what error you're getting - it's hard to understand from the above.

**Comment by slava-se** [comment]
@roji 

While executing ExecuteUpdateAsync() I'm trying to set 2 columns on every call and on every entity which implements Auditable 
1) DateTime DateUpdated column which works with the code from this [comment](https://github.com/dotnet/efcore/issues/35105#issuecomment-2486311979)

2) BrowserInfo jsonB column with the browser's information of the requesting user which I'm taking from the context -> **that's where I need help**. The code mentioned in 1) seems to work with CLR types only and not with custom types. But the implementation for CLR types also looks rather complicated. 
Is there a way to set data of non-CLR type during ExecuteUpdateAsync() automatically?


```
public class User : Auditable
{
    [Key]
    public int Id { get; set; }
    public string Name { get; set; }
}

public abstract class Auditable
{
	[DatabaseGenerated(DatabaseGeneratedOption.Identity)]
	public DateTime DateCreated { get; set; }

	public DateTime DateUpdated { get; set; }

	[Column(TypeName = "jsonb")]
	public BrowserInfo? BrowserInfo { get; set; }
}

public class BrowserInfo
{
    [JsonConverter(typeof(JsonStringEnumConverter))]
    public Browser Browser { get; init; }
    public Version BrowserVersion { get; init; } = null!;
}
```

**Comment by roji** [comment]
@slava-se can you please put together of a simple, minimal repro using ExecuteUpdateAsync directly (no dynamic expression trees) to show exactly what you're trying to do? Please also post the exception.

**Comment by slava-se** [comment]
@roji by saying repro, do you mean a runnable project as an archive?

to simplify my question, I have many entities inheriting from Auditable which I've mentioned above, e.g
```
public class User : Auditable
{
    [Key]
    public int Id { get; set; }
    public string Name { get; set; }
}

public class UserAddress : Auditable
{
    [Key]
    public int Id { get; set; }
    public int UserId { get; set; }
    public string FullAddress { get; set; }
}
```

now, if I'm updating User and UserAddress  and willing to update the Auditable.DateUpdated and Auditable.BrowserInfo I would need to call

`await context.Users.Where(u => u.Id == 1).ExecuteUpdateAsync(u => u.SetProperty(p => p.Name, "newName").SetProperty(p => p.DateUpdated, DateTime.UtcNow).SetProperty(p => p.BrowserInfo = browserInfoFromContext));`

`await context.UserAddresses.Where(u => u.UserId == 1).ExecuteUpdateAsync(u => u.SetProperty(p => p.FullAddress, "newAddress").SetProperty(p => p.DateUpdated, DateTime.UtcNow).SetProperty(p => p.BrowserInfo = browserInfoFromContext));`

So what I'm looking for is some kind of middleware or interceptor which will check if the entity I'm executing ExecuteUpdateAsync() on is an auditable and will set its DateUpdated and BrowserInfo there so I don't have to do it in every single repository.

So basically I'm looking for a way to call
`await context.Users.Where(u => u.Id == 1).ExecuteUpdateAsync(u => u.SetProperty(p => p.Name, "newName"));`
without explicitely setting DateUpdated and BrowserInfo every single time and having an interceptor which will set DateUpdated and BrowserInfo in the background

**Comment by roji** [comment]
> by saying repro, do you mean a runnable project as an archive?

In general yes - please do that whenever you submit an issue, as long textual explanations seldom successfully convey what it is you're trying to do, but a simple code sample does.

> So what I'm looking for is some kind of middleware or interceptor which will check if the entity I'm executing ExecuteUpdateAsync() on is an auditable and will set its DateUpdated and BrowserInfo there so I don't have to do it in every single repository.

The closest thing we have to this is [LINQ query tree interception](https://learn.microsoft.com/en-us/ef/core/what-is-new/ef-core-7.0/whatsnew#linq-expression-tree-interception): this allows you to intercept the LINQ query tree early inside EF. You could write code which identifies ExecuteUpdateAsync, checks the type, and then manipulates the setters to add new setters. This is likely not going to be trivial, but should be doable.

Note that because of #32018, you'll likely need to change this logic a bit for EF 10, but the general idea will still work.

I'll go ahead and close this the answer has been answered, but if you have further questions/comments feel free to continue posting.
